stdlib/public/runtime/SwiftValue.mm

```diff
@@ -63,65 +64,151 @@ - (id)copyWithZone:(NSZone *)zone;
 
   /// The witness table for `Hashable` conformance.
   /// This member is only available for native Swift errors.
   /// This member is lazily-initialized.
   /// Instead of using it directly, call `getHashableConformance()`.
-  mutable std::atomic<const hashable_support::HashableWitnessTable *>
-      hashableConformance;
+  mutable std::atomic<const void *> cachedConformance;
 
   /// Get the base type that conforms to `Hashable`.
   /// Returns NULL if the type does not conform.
   const Metadata *getHashableBaseType() const;
 
+  /// Get the base type that conforms to `Equatable`.
+  /// Returns NULL if the type does not conform.
+  const Metadata *getEquatableBaseType() const;
+
   /// Get the `Hashable` protocol witness table for the contained type.
   /// Returns NULL if the type does not conform.
   const hashable_support::HashableWitnessTable *getHashableConformance() const;
 
+  /// Get the `Equatable` protocol witness table for the contained type.
+  /// Returns NULL if the type does not conform.
+  const equatable_support::EquatableWitnessTable *getEquatableConformance() const;
+
+  /// Populate the `cachedConformance` with the Hashable conformance
+  /// (if there is one), else the Equatable conformance.
+  const void * cacheHashableEquatableConformance() const;
+
+
   SwiftValueHeader()
-      : hashableBaseType(nullptr), hashableConformance(nullptr) {}
+      : cachedBaseType(nullptr), cachedConformance((void *)nullptr) {}
 };
 
 const Metadata *SwiftValueHeader::getHashableBaseType() const {
-  if (auto type = hashableBaseType.load(std::memory_order_acquire)) {
-    if (reinterpret_cast<uintptr_t>(type) == 1) {
-      return nullptr;
-    }
+  auto type = cachedBaseType.load(std::memory_order_acquire);
+  if (type == nullptr) {
+    cacheHashableEquatableConformance();
+    type = cachedBaseType.load(std::memory_order_acquire);
+  }
+  if ((reinterpret_cast<uintptr_t>(type) & 1) == 0) {
     return type;
+  } else {
+    return nullptr;
   }
+}
+
+const Metadata *SwiftValueHeader::getEquatableBaseType() const {
+  auto type = cachedBaseType.load(std::memory_order_acquire);
+  if (type == nullptr) {
+    cacheHashableEquatableConformance();
+    type = cachedBaseType.load(std::memory_order_acquire);
+  }
+  if ((reinterpret_cast<uintptr_t>(type) & 1) == 0) {
+    // A Hashable conformance was found
+    return nullptr;
+  } else {
+    // An Equatable conformance (or neither) was found
+    return reinterpret_cast<const Metadata *>(reinterpret_cast<uintptr_t>(type) & ~1ULL);
+  }
+}
+
+// Set cachedConformance to the Hashable conformance if
+// there is one, else the Equatable conformance.
+// Also set cachedBaseType to the parent type that
+// introduced the Hashable/Equatable conformance.
+// The cached conformance and type set the LSbit to indicate
+// which conformance is present:
+// * If the LSbit is not set, it's the Hashable conformance
+// * If the value is exactly 1, neither conformance is present
```

Comment:
Nice way to cache the negative lookup. I was about to suggest doing this, then saw you already did it.